datafusion/core/src/datasource/file_format/parquet.rs

```
@@ -828,113 +833,84 @@ impl DataSink for ParquetSink {
-    let mut join_handles: Vec<JoinHandle<ParquetFileSerializedResult>> =
-        Vec::with_capacity(parallelism);
-    for _ in 0..parallelism {
-        let buffer: Vec<u8> = Vec::new();
-        let mut writer = parquet::arrow::arrow_writer::ArrowWriter::try_new(
-            buffer,
-            output_schema.clone(),
-            Some(parquet_props.clone()),
-        )?;
-        let mut data_stream = data.remove(0);
-        join_handles.push(tokio::spawn(async move {
-            let mut inner_row_count = 0;
-            while let Some(batch) = data_stream.next().await.transpose()? {
-                inner_row_count += batch.num_rows();
-                writer.write(&batch)?;
+    let (serialize_tx, mut serialize_rx) =
+        mpsc::channel::<JoinHandle<RBStreamSerializeResult>>(100);
+
+    // Create some Arc<> copies that we can move into launch_serialization task and still access later
+    let arc_props = Arc::new(parquet_props.clone());
+    let arc_props_clone = arc_props.clone();
+    let schema_clone = output_schema.clone();
+    let launch_serialization_task: JoinHandle<Result<(), DataFusionError>> =
+        tokio::spawn(async move {
+            for mut stream in data {
+                let schema_desc = arrow_to_parquet_schema(&schema_clone)?;
+                let mut writer = ArrowRowGroupWriter::new(
+                    &schema_desc,
+                    &arc_props_clone,
+                    &schema_clone,
+                )?;
+                serialize_tx
+                    .send(tokio::spawn(async move {
+                        let mut inner_row_count = 0;
+                        while let Some(rb) = stream.next().await.transpose()? {
+                            inner_row_count += rb.num_rows();
+                            writer.write(&rb)?;
+                        }
+                        Ok((writer.close()?, inner_row_count))
+                    }))
+                    .await
+                    .map_err(|_| {
```

I think the only way the send will fail is if the receiver was dropped -- either due to early plan cancel or some other error

Thus it might make sense to ignore the error and return (with a comment about rationale) rather than returning an error